README.md
# Naive benchmarking...

Made the main benchmark code single-threaded.
Also changed `go` calls for `thread` calls, which solved the slow reads issue (not enough threads in the pool when using `go` and `go-loop`8-11-21 19:39:50 Lambdabeast2 INFO [redisbench.core:85] - Added 31599, read 222486, deleted 8 - 269545 keys in db
18-11-21 19:39:51 Lambdabeast2 INFO [redisbench.core:85] - Added 29896, read 224814, deleted 8 - 282911 keys in db
18-11-21 19:39:52 Lambdabeast2 INFO [redisbench.core:85] - Added 32175, read 214360, deleted 8 - 296616 keys in db
18-11-21 19:39:53 Lambdabeast2 INFO [redisbench.core:85] - Added 32328, read 218948, deleted 7 - 309273 keys in db
18-11-21 19:39:54 Lambdabeast2 INFO [redisbench.core:85] - Added 32047, read 217592, deleted 7 - 321201 keys in db
8-11-21 19:39:50 Lambdabeast2 INFO [redisbench.core:85] - Added 31599, read 222486, deleted 8 - 269545 keys in db
18-11-21 19:39:51 Lambdabeast2 INFO [redisbench.core:85] - Added 29896, read 224814, deleted 8 - 282911 keys in db
18-11-21 19:39:52 Lambdabeast2 INFO [redisbench.core:85] - Added 32175, read 214360, deleted 8 - 296616 keys in db
18-11-21 19:39:53 Lambdabeast2 INFO [redisbench.core:85] - Added 32328, read 218948, deleted 7 - 309273 keys in db
18-11-21 19:39:54 Lambdabeast2 INFO [redisbench.core:85] - Added 32047, read 217592, deleted 7 - 321201 keys in db
)
Added criterium based benchmarking for the reads, and fixed payload size bug.


## 500K keys, 360B each, with 10KB payload
```
λ lein run
18-11-15 22:22:25 Lambdabeast2 INFO [redisbench.core:121] - Starting...
18-11-15 22:22:25 Lambdabeast2 INFO [redisbench.core:122] - Inserting 500000 keys with 10000 bytes each...
18-11-15 22:23:18 Lambdabeast2 INFO [redisbench.core:128] - Inserts took 53.561s (0.107122ms per insert)
18-11-15 22:23:18 Lambdabeast2 INFO [redisbench.core:129] - Starting read benchmark...
Estimating sampling overhead
Warming up for JIT optimisations 10000000000 ...
  compilation occurred before 830336 iterations
  compilation occurred before 107086464 iterations
  compilation occurred before 209191962 iterations
  compilation occurred before 211682340 iterations
  compilation occurred before 317938468 iterations
  compilation occurred before 424194596 iterations
  classes loaded before 670742018 iterations
Estimating execution count ...
Sampling ...
Final GC...
Checking GC...
Finding outliers ...
Bootstrapping ...
Checking outlier significance
Warming up for JIT optimisations 5000000000 ...
  compilation occurred before 6078 iterations
Estimating execution count ...
Sampling ...
Final GC...
Checking GC...
Finding outliers ...
Bootstrapping ...
Checking outlier significance
Evaluation count : 7278 in 6 samples of 1213 calls.
             Execution time mean : 76.761927 µs
    Execution time std-deviation : 6.446678 µs
   Execution time lower quantile : 70.239845 µs ( 2.5%)
   Execution time upper quantile : 83.911379 µs (97.5%)
                   Overhead used : 9.017062 ns
```

## 500K keys, 360B each, with 1KB payload
```
λ lein run
18-11-15 22:30:08 Lambdabeast2 INFO [redisbench.core:121] - Starting...
18-11-15 22:30:08 Lambdabeast2 INFO [redisbench.core:122] - Inserting 500000 keys with 1000 bytes each...
18-11-15 22:30:42 Lambdabeast2 INFO [redisbench.core:128] - Inserts took 33.79s (0.06758ms per insert)
18-11-15 22:30:42 Lambdabeast2 INFO [redisbench.core:129] - Starting read benchmark...
Estimating sampling overhead
Warming up for JIT optimisations 10000000000 ...
  compilation occurred before 104950467 iterations
  compilation occurred before 109182332 iterations
  compilation occurred before 214978957 iterations
  compilation occurred before 324161074 iterations
  compilation occurred before 432496818 iterations
Estimating execution count ...
Sampling ...
Final GC...
Checking GC...
Finding outliers ...
Bootstrapping ...
Checking outlier significance
Warming up for JIT optimisations 5000000000 ...
  compilation occurred before 9768 iterations
Estimating execution count ...
Sampling ...
Final GC...
Checking GC...
Finding outliers ...
Bootstrapping ...
Checking outlier significance
Evaluation count : 12402 in 6 samples of 2067 calls.
             Execution time mean : 49.882914 µs
    Execution time std-deviation : 12.579224 µs
   Execution time lower quantile : 43.662113 µs ( 2.5%)
   Execution time upper quantile : 71.496311 µs (97.5%)
                   Overhead used : 8.851501 ns

Found 1 outliers in 6 samples (16.6667 %)
        low-severe       1 (16.6667 %)
 Variance from outliers : 64.8466 % Variance is severely inflated by outliers
```

## 500K keys, 720B each, with 1KB payload 
```
λ lein run
18-11-15 22:38:49 Lambdabeast2 INFO [redisbench.core:131] - Starting...
18-11-15 22:38:49 Lambdabeast2 INFO [redisbench.core:132] - Inserting 500000 keys with 1000 bytes each...
18-11-15 22:39:28 Lambdabeast2 INFO [redisbench.core:138] - Inserts took 38.983s (0.077966ms per insert)
18-11-15 22:39:28 Lambdabeast2 INFO [redisbench.core:140] - Reading 500000 keys with 1000 bytes each...
18-11-15 22:39:52 Lambdabeast2 INFO [redisbench.core:146] - Reads took 24.157s (0.048314ms per read)
18-11-15 22:39:52 Lambdabeast2 INFO [redisbench.core:147] - Starting read benchmark...
Estimating sampling overhead
Warming up for JIT optimisations 10000000000 ...
  compilation occurred before 824585 iterations
  compilation occurred before 102222464 iterations
  compilation occurred before 106344329 iterations
  compilation occurred before 209390954 iterations
  compilation occurred before 220932176 iterations
  compilation occurred before 421254815 iterations
Estimating execution count ...
Sampling ...
Final GC...
Checking GC...
Finding outliers ...
Bootstrapping ...
Checking outlier significance
Warming up for JIT optimisations 5000000000 ...
  compilation occurred before 4821 iterations
Estimating execution count ...
Sampling ...
Final GC...
Checking GC...
Finding outliers ...
Bootstrapping ...
Checking outlier significance
Evaluation count : 8856 in 6 samples of 1476 calls.
             Execution time mean : 52.494473 µs
    Execution time std-deviation : 4.733230 µs
   Execution time lower quantile : 47.353977 µs ( 2.5%)
   Execution time upper quantile : 57.939047 µs (97.5%)
                   Overhead used : 8.905395 ns
18-11-15 22:40:19 Lambdabeast2 INFO [redisbench.core:150] - Starting write benchmark...
Warming up for JIT optimisations 5000000000 ...
  compilation occurred before 8295 iterations
Estimating execution count ...
Sampling ...
Final GC...
Checking GC...
Finding outliers ...
Bootstrapping ...
Checking outlier significance
Evaluation count : 8472 in 6 samples of 1412 calls.
             Execution time mean : 80.945854 µs
    Execution time std-deviation : 6.181974 µs
   Execution time lower quantile : 76.103014 µs ( 2.5%)
   Execution time upper quantile : 88.479525 µs (97.5%)
                   Overhead used : 8.905395 ns
18-11-15 22:40:27 Lambdabeast2 INFO [redisbench.core:154] - Deleting 500000 keys with 1000 bytes each...
18-11-15 22:40:53 Lambdabeast2 INFO [redisbench.core:160] - Deletes took 25.971s (0.051942ms per delete)
```

## 500K keys, 720B each, with 10KB payload (redis dump to disk disabled)
```
λ lein run
18-11-15 23:04:40 Lambdabeast2 INFO [redisbench.core:131] - Starting...
18-11-15 23:04:40 Lambdabeast2 INFO [redisbench.core:132] - Inserting 500000 keys with 10000 bytes each...
18-11-15 23:05:16 Lambdabeast2 INFO [redisbench.core:138] - Inserts took 36.421s (0.072842ms per insert)
18-11-15 23:05:16 Lambdabeast2 INFO [redisbench.core:140] - Reading 500000 keys with 10000 bytes each...
18-11-15 23:05:41 Lambdabeast2 INFO [redisbench.core:146] - Reads took 24.871s (0.049742ms per read)
18-11-15 23:05:41 Lambdabeast2 INFO [redisbench.core:147] - Starting read benchmark...
Estimating sampling overhead
Warming up for JIT optimisations 10000000000 ...
  compilation occurred before 790213 iterations
  compilation occurred before 101911621 iterations
  compilation occurred before 199872985 iterations
  compilation occurred before 302574415 iterations
  compilation occurred before 403695823 iterations
Estimating execution count ...
Sampling ...
Final GC...
Checking GC...
Finding outliers ...
Bootstrapping ...
Checking outlier significance
Warming up for JIT optimisations 5000000000 ...
  compilation occurred before 6544 iterations
Estimating execution count ...
Sampling ...
Final GC...
Checking GC...
Finding outliers ...
Bootstrapping ...
Checking outlier significance
Evaluation count : 11718 in 6 samples of 1953 calls.
             Execution time mean : 51.909264 µs
    Execution time std-deviation : 1.359535 µs
   Execution time lower quantile : 50.519779 µs ( 2.5%)
   Execution time upper quantile : 53.528169 µs (97.5%)
                   Overhead used : 8.802747 ns
18-11-15 23:06:09 Lambdabeast2 INFO [redisbench.core:150] - Starting write benchmark...
Warming up for JIT optimisations 5000000000 ...
  compilation occurred before 7683 iterations
Estimating execution count ...
Sampling ...
Final GC...
Checking GC...
Finding outliers ...
Bootstrapping ...
Checking outlier significance
Evaluation count : 11550 in 6 samples of 1925 calls.
             Execution time mean : 53.696089 µs
    Execution time std-deviation : 3.049825 µs
   Execution time lower quantile : 50.891543 µs ( 2.5%)
   Execution time upper quantile : 58.163106 µs (97.5%)
                   Overhead used : 8.802747 ns
18-11-15 23:06:18 Lambdabeast2 INFO [redisbench.core:154] - Deleting 500000 keys with 10000 bytes each...
18-11-15 23:06:42 Lambdabeast2 INFO [redisbench.core:160] - Deletes took 23.698s (0.047396ms per delete)
```

## 500K keys, 1440B each, with 500B payload (redis dump to disk disabled)
```
λ lein run
18-11-15 23:09:53 Lambdabeast2 INFO [redisbench.core:131] - Starting...
18-11-15 23:09:53 Lambdabeast2 INFO [redisbench.core:132] - Inserting 500000 keys with 500 bytes each...
18-11-15 23:10:20 Lambdabeast2 INFO [redisbench.core:138] - Inserts took 27.053s (0.054106ms per insert)
18-11-15 23:10:20 Lambdabeast2 INFO [redisbench.core:140] - Reading 500000 keys with 500 bytes each...
18-11-15 23:10:44 Lambdabeast2 INFO [redisbench.core:146] - Reads took 24.142s (0.048284ms per read)
18-11-15 23:10:44 Lambdabeast2 INFO [redisbench.core:147] - Starting read benchmark...
Estimating sampling overhead
Warming up for JIT optimisations 10000000000 ...
  compilation occurred before 877797 iterations
  compilation occurred before 113206629 iterations
  compilation occurred before 222025185 iterations
  compilation occurred before 232556013 iterations
  compilation occurred before 448437987 iterations
  classes loaded before 923202816 iterations
  compilation occurred before 924080385 iterations
Estimating execution count ...
Sampling ...
Final GC...
Checking GC...
Finding outliers ...
Bootstrapping ...
Checking outlier significance
Warming up for JIT optimisations 5000000000 ...
  compilation occurred before 7734 iterations
Estimating execution count ...
Sampling ...
Final GC...
Checking GC...
Finding outliers ...
Bootstrapping ...
Checking outlier significance
Evaluation count : 12726 in 6 samples of 2121 calls.
             Execution time mean : 50.424129 µs
    Execution time std-deviation : 3.945189 µs
   Execution time lower quantile : 47.571508 µs ( 2.5%)
   Execution time upper quantile : 57.027016 µs (97.5%)
                   Overhead used : 8.857620 ns

Found 1 outliers in 6 samples (16.6667 %)
        low-severe       1 (16.6667 %)
 Variance from outliers : 15.4850 % Variance is moderately inflated by outliers
18-11-15 23:11:12 Lambdabeast2 INFO [redisbench.core:150] - Starting write benchmark...
Warming up for JIT optimisations 5000000000 ...
  compilation occurred before 8782 iterations
Estimating execution count ...
Sampling ...
Final GC...
Checking GC...
Finding outliers ...
Bootstrapping ...
Checking outlier significance
Evaluation count : 12078 in 6 samples of 2013 calls.
             Execution time mean : 55.930212 µs
    Execution time std-deviation : 5.130326 µs
   Execution time lower quantile : 51.215809 µs ( 2.5%)
   Execution time upper quantile : 62.731403 µs (97.5%)
                   Overhead used : 8.857620 ns
18-11-15 23:11:21 Lambdabeast2 INFO [redisbench.core:154] - Deleting 500000 keys with 500 bytes each...
18-11-15 23:11:44 Lambdabeast2 INFO [redisbench.core:160] - Deletes took 22.859s (0.045718ms per delete)
```


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2018-11-15
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2018-11-15
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/redisbench/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/redisbench/compare/0.1.0...0.1.1


test/redisbench/core_test.clj
(ns redisbench.core-test
  (:require [clojure.test :refer :all]
            [redisbench.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/redisbench/core.clj
(ns redisbench.core
  (:gen-class)
  (:require [clojure.core.async :as async :refer [>! <! >!! <!! go go-loop thread chan close!]]
            [taoensso.timbre :as log]
            [taoensso.carmine :as car]
            [criterium.core :as criterium]
            [clojurewerkz.spyglass.client :as c]
            [clojure.java.io :as io :refer [file  input-stream]]))


;; redis stuff
(def server1-conn {:spec {:redis-host "localhost" :redis-port 6379}}) ; See `wcar` docstring for opts
(defmacro wcar* [& body] `(car/wcar server1-conn ~@body))

;;memcached stuff
;;(def tmc (c/bin-connection "127.0.0.1:11211"))


(defn uuid [] (str (java.util.UUID/randomUUID)))



(defn gen-keys
  "Create a vector of `n` keys, each keys being a `mult` UUIDs concatenated"
  [n mult]
  {:pre [(pos? mult)
         (pos? n)]}
  (letfn [(f []
            (apply str (take mult (repeatedly uuid))))]
    (vec (take n (repeatedly f)))))


(defn gen-payloads
  "Create a vector of `n` random chunks of data, each `len` bytes long"
  [n len]
  {:pre [(pos? len)
         (pos? n)]}
  (with-open [in (input-stream (file "/dev/urandom"))]
    (let [read-chunk #(let [buf (byte-array len)]
                        (.read in buf)
                        buf)]
      (vec (take n (repeatedly read-chunk))))))

(defn launch-writer
  [channel keys-source payloads-source add-counter-atom]
  (thread (while channel
        (let [k (rand-nth keys-source)
              p (rand-nth payloads-source)]
          (wcar* (car/set k p))
          ;(c/set tmc k 3600 p)
          (swap! add-counter-atom inc)
          (>!! channel k)))))

(defn launch-updater
  [channel keys-atom]
  (thread (loop []
            (let [k (<!! channel)]
              (swap! keys-atom conj k)
              (recur)))))

(defn launch-remover
  [keys-atom del-counter-atom continue?]
  (thread (do
        (Thread/sleep 1000)
        (while @continue?
          (let [k (rand-nth (vec @keys-atom))]
            (when k
              (do
                  ;(c/delete tmc k)
                  (wcar* (car/unlink k))
                  (swap! del-counter-atom inc)
                  (swap! keys-atom disj k)))
            (Thread/sleep 20))))))



(defn launch-logger
  [add-counter read-counter del-counter continue?]
  (thread
    (while @continue?
      (let [dbsize  #_(-> (c/get-stats tmc)
                       vals
                       (#(into {} %))
                       (get "total_items"))  (wcar* (car/dbsize))]
        (log/info (str "Added " @add-counter
                       ", read " @read-counter
                       ", deleted " @del-counter
                       " - " dbsize " keys in db"))
        (reset! add-counter 0)
        (reset! del-counter 0)
        (reset! read-counter 0)
        (Thread/sleep 1000)))))

(defn launch-reader
  [keys-source read-atom continue?]
  (thread (do 
            (Thread/sleep 500)
            (while @continue?
              (when-let [key (rand-nth keys-source)]
                (wcar* car/get key)
                ;; (c/get tmc key)
                (swap! read-atom inc))))))

(defn put-all-keys [ks vs]
  (doseq [k ks]
    (wcar* (car/set k (rand-nth vs)))))


(defn read-all-keys [ks]
  (doseq [k ks]
    (wcar* (car/get k))))


(defn delete-all-keys [ks]
  (doseq [k ks]
    (wcar* (car/unlink k))))


(defn -main
  [& args]
  (let [keys-source (gen-keys 500000 40)
        payloads-source (gen-payloads 1000 500)
        added-keys (atom #{})
        add-counter (atom 0)
        del-counter (atom 0)
        read-counter (atom 0)
        timer (atom 0)
        continue? (atom true)
        channel (chan 100)]
    (wcar* (car/flushall))
    (log/info "Starting...")
    (log/info (str "Inserting " (count keys-source) " keys with "
                   (count (first payloads-source)) " bytes each..."))
    (reset! timer (System/currentTimeMillis))
    (put-all-keys keys-source payloads-source)
    (let [time-consumed (- (System/currentTimeMillis) @timer)
          avg (float (/ time-consumed (count keys-source)))]
      (log/info (str "Inserts took " (double (/ time-consumed 1000)) "s (" avg "ms per insert)")))

    (log/info (str "Reading " (count keys-source) " keys with "
                   (count (first payloads-source)) " bytes each..."))
    (reset! timer (System/currentTimeMillis))
    (read-all-keys keys-source)
    (let [time-consumed (- (System/currentTimeMillis) @timer)
          avg (float (/ time-consumed (count keys-source)))]
      (log/info (str "Reads took " (double (/ time-consumed 1000)) "s (" avg "ms per read)")))
    (log/info "Starting read benchmark...")
    (criterium/with-progress-reporting
      (criterium/quick-bench (wcar* (car/get (rand-nth keys-source)))))
    (log/info "Starting write benchmark...")
    (criterium/with-progress-reporting
      (criterium/quick-bench (wcar* (car/set (rand-nth keys-source) (rand-nth payloads-source)))))

    (log/info (str "Deleting " (count keys-source) " keys with "
                   (count (first payloads-source)) " bytes each..."))
    (reset! timer (System/currentTimeMillis))
    (delete-all-keys keys-source)
    (let [time-consumed (- (System/currentTimeMillis) @timer)
          avg (float (/ time-consumed (count keys-source)))]
      (log/info (str "Deletes took " (double (/ time-consumed 1000)) "s (" avg "ms per delete)")))

    ;; (dotimes [n 4](launch-writer channel keys-source payloads-source add-counter))
    ;; (launch-updater channel added-keys)
    ;; (launch-remover added-keys del-counter continue?)
    ;; (dotimes [n 3] (launch-reader keys-source read-counter continue?))
    ;; (Thread/sleep 10000)
    ;; (launch-logger add-counter read-counter del-counter continue?)
    (Thread/sleep 300000)
    (shutdown-agents)
    (wcar* (car/flushall))))


